File: buildSrc/src/main/java/elfeatures/gradle/tasks/TransformJarContentCopyAction.java
package elfeatures.gradle.tasks;

import lombok.AllArgsConstructor;
import org.apache.commons.compress.archivers.jar.JarArchiveEntry;
import org.apache.commons.compress.archivers.jar.JarArchiveOutputStream;
import org.apache.commons.compress.archivers.zip.UnixStat;
import org.gradle.api.Action;
import org.gradle.api.GradleException;
import org.gradle.api.file.FileCopyDetails;
import org.gradle.api.internal.file.CopyActionProcessingStreamAction;
import org.gradle.api.internal.file.copy.CopyAction;
import org.gradle.api.internal.file.copy.CopyActionProcessingStream;
import org.gradle.api.internal.file.copy.FileCopyDetailsInternal;
import org.gradle.api.tasks.WorkResult;
import org.gradle.api.tasks.WorkResults;
import org.gradle.internal.IoActions;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.function.UnaryOperator;

@AllArgsConstructor
public class TransformJarContentCopyAction implements CopyAction {

    private final Path jarFile;
    private final UnaryOperator<String> fileEntryNameTransformer;
    private final UnaryOperator<String> dirEntryNameTransformer;

    @Override
    public WorkResult execute(CopyActionProcessingStream stream) {
        final JarArchiveOutputStream zipOutStr;

        try {
            if (!Files.isDirectory(jarFile.getParent()))
                Files.createDirectories(jarFile.getParent());

            zipOutStr = new JarArchiveOutputStream(Files.newOutputStream(jarFile, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING));
        } catch (Exception ex) {
            throw new GradleException(String.format("Could not create JAR '%s'.", jarFile), ex);
        }

        try {
            Action<? super JarArchiveOutputStream> streamAction = outputStream -> stream.process(new StreamAction(outputStream));
            IoActions.withResource(zipOutStr, streamAction);
        } catch (Exception ex) {
            try {
                Files.deleteIfExists(jarFile);
            } catch (IOException ignored) {
            }
            throw ex;
        }

        return WorkResults.didWork(true);
    }

    @AllArgsConstructor
    private class StreamAction implements CopyActionProcessingStreamAction {

        private final JarArchiveOutputStream jarOutStr;

        @Override
        public void processFile(FileCopyDetailsInternal details) {
            if (details.isDirectory()) {
                visitDir(details);
            } else {
                visitFile(details);
            }
        }

        private void visitFile(FileCopyDetails fileDetails) {
            try {
                String entryName = fileEntryNameTransformer.apply(fileDetails.getRelativePath().getPathString());
                JarArchiveEntry archiveEntry = new JarArchiveEntry(entryName);
                archiveEntry.setTime(fileDetails.getLastModified());
                archiveEntry.setUnixMode(UnixStat.FILE_FLAG | fileDetails.getPermissions().toUnixNumeric());
                jarOutStr.putArchiveEntry(archiveEntry);
                fileDetails.copyTo(jarOutStr);
                jarOutStr.closeArchiveEntry();
            } catch (Exception ex) {
                throw new GradleException(String.format("Could not add %s to JAR '%s'.", fileDetails, jarFile), ex);
            }
        }

        private void visitDir(FileCopyDetails dirDetails) {
            try {
                String entryName = dirEntryNameTransformer.apply(dirDetails.getRelativePath().getPathString() + '/');
                JarArchiveEntry archiveEntry = new JarArchiveEntry(entryName);
                archiveEntry.setTime(dirDetails.getLastModified());
                archiveEntry.setUnixMode(UnixStat.DIR_FLAG | dirDetails.getPermissions().toUnixNumeric());
                jarOutStr.putArchiveEntry(archiveEntry);
                jarOutStr.closeArchiveEntry();
            } catch (Exception ex) {
                throw new GradleException(String.format("Could not add %s to JAR '%s'.", dirDetails, jarFile), ex);
            }
        }

    }

}

File: buildSrc/src/main/kotlin/elfeatures/gradle/task/TransformJarContentCopyAction.kt
package elfeatures.gradle.task

import org.apache.commons.compress.archivers.jar.JarArchiveEntry
import org.apache.commons.compress.archivers.jar.JarArchiveOutputStream
import org.apache.commons.compress.archivers.zip.UnixStat
import org.gradle.api.GradleException
import org.gradle.api.file.FileCopyDetails
import org.gradle.api.internal.file.CopyActionProcessingStreamAction
import org.gradle.api.internal.file.copy.CopyAction
import org.gradle.api.internal.file.copy.CopyActionProcessingStream
import org.gradle.api.internal.file.copy.FileCopyDetailsInternal
import org.gradle.api.tasks.WorkResult
import org.gradle.api.tasks.WorkResults
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardOpenOption
import java.util.function.UnaryOperator

class TransformJarContentCopyAction(
    private val jarFile: Path,
    private val fileEntryNameTransformer: UnaryOperator<String>,
    private val dirEntryNameTransformer: UnaryOperator<String>
) : CopyAction {

    override fun execute(stream: CopyActionProcessingStream): WorkResult {
        val zipOutStr: JarArchiveOutputStream

        try {
            if (!Files.isDirectory(jarFile.parent))
                Files.createDirectories(jarFile.parent)

            zipOutStr = JarArchiveOutputStream(
                Files.newOutputStream(
                    jarFile,
                    StandardOpenOption.CREATE,
                    StandardOpenOption.TRUNCATE_EXISTING
                )
            )
        } catch (ex: Exception) {
            throw GradleException("Couldn't create JAR '$jarFile'", ex)
        }

        try {
            zipOutStr.use { stream.process(StreamAction(it)) }
        } catch (ex: Exception) {
            try {
                Files.deleteIfExists(jarFile)
            } catch (_: IOException) {
            }
            throw ex
        }

        return WorkResults.didWork(true)
    }

    private inner class StreamAction(
        private val jarOutStr: JarArchiveOutputStream
    ) : CopyActionProcessingStreamAction {

        override fun processFile(details: FileCopyDetailsInternal) {
            if (details.isDirectory) {
                visitDir(details)
            } else {
                visitFile(details)
            }
        }

        fun visitFile(fileDetails: FileCopyDetails) {
            try {
                val entryName = fileEntryNameTransformer.apply(fileDetails.relativePath.getPathString())
                val archiveEntry = JarArchiveEntry(entryName)
                archiveEntry.time = fileDetails.lastModified
                archiveEntry.setUnixMode(UnixStat.FILE_FLAG or fileDetails.permissions.toUnixNumeric())
                jarOutStr.putArchiveEntry(archiveEntry)
                fileDetails.copyTo(jarOutStr)
                jarOutStr.closeArchiveEntry()
            } catch (ex: Exception) {
                throw GradleException("Couldn't add $fileDetails to JAR '$jarFile'", ex)
            }
        }

        fun visitDir(dirDetails: FileCopyDetails) {
            try {
                val entryName = dirEntryNameTransformer.apply(dirDetails.relativePath.getPathString() + '/')
                val archiveEntry = JarArchiveEntry(entryName)
                archiveEntry.time = dirDetails.lastModified
                archiveEntry.setUnixMode(UnixStat.DIR_FLAG or dirDetails.permissions.toUnixNumeric())
                jarOutStr.putArchiveEntry(archiveEntry)
                jarOutStr.closeArchiveEntry()
            } catch (ex: Exception) {
                throw GradleException("Couldn't add $dirDetails to JAR '$jarFile'", ex)
            }
        }
    }
}